### ISSUE TITLE
[MINSTALL-151] Projects without primary artifacts, but with attachment artifacts fail install

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-install-plugin#313]
**[Peter Huffer](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=peter.huffer)** opened **[MINSTALL-151](https://issues.apache.org/jira/browse/MINSTALL-151?redirect=false)** and commented

I have a module which has a packaging of `feature` from the `karaf-maven-plugin`. When running the `mvn install` command, the following error occurs:

```java
Failed to execute goal org.apache.maven.plugins:maven-install-plugin:3.0.0-M1:install (default-install) on project <project_name>: NoFileAssignedException: The packaging plugin for this project did not assign a main file to the project but it has attachments. Change packaging to 'pom'. -> [Help 1]
```

Reverting back to version `2.5.2` made the `mvn install` command succeed.

---

**Affects:** 3.0.0-M1

**Attachments:**
- [minstall-151.zip](https://issues.apache.org/jira/secure/attachment/12941997/minstall-151.zip) (_9.04 kB_)
- [minstall-151-example-PeterSchloemer.zip](https://issues.apache.org/jira/secure/attachment/13047315/minstall-151-example-PeterSchloemer.zip) (_1.17 kB_)
- [mvn-install-fail.zip](https://issues.apache.org/jira/secure/attachment/13056525/mvn-install-fail.zip) (_5.57 kB_)

**Issue Links:**
- [MSHARED-760](https://issues.apache.org/jira/browse/MSHARED-760) Do not fail if no main artifact has been set, but attachments existing
 (_**"is blocked by"**_)
- [MINSTALL-189](https://issues.apache.org/jira/browse/MINSTALL-189) Add parameter to lax project validation
 (_**"is fixed by"**_)
- [MINSTALL-152](https://issues.apache.org/jira/browse/MINSTALL-152) Maven Install Plugin Version 3.0.0-M1 unable to build POMs with packaging set "feature"
 (_**"duplicates"**_)
- [MDEPLOY-288](https://issues.apache.org/jira/browse/MDEPLOY-288) Deploy is failed when no primary artifact is present

- [MINSTALL-118](https://issues.apache.org/jira/browse/MINSTALL-118) MavenProject with only attachments must have packaging "pom"

- [MINSTALL-169](https://issues.apache.org/jira/browse/MINSTALL-169) Handle non-pom packaging with only a pom artifact

7 votes, 29 watchers

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Can you please attach a full working example ?

**Comment by jira-importer** [comment]
**[Charles Karney](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cffk)** commented

This is a show-stopper for us.  Our entire build framework is down.  Is it possible to change pom.xml to get it to use maven-install-plugin 2.5.2.

**Comment by jira-importer** [comment]
**[Peter Huffer](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=peter.huffer)** commented

[Charles Karney](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cffk) In our root pom in the pluginManagement section we did 

```java
<!-- Override version for maven-install-plugin because there is a bug in
 3.0.0-M1 preventing installing of modules with packaging of feature
 see: https://issues.apache.org/jira/browse/MINSTALL-151 -->
<plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-install-plugin</artifactId>
    <version>2.5.2</version>
</plugin>
```

as a fix for now. It worked for us.

 

[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise) Yup, I'll work on a quick example.

**Comment by jira-importer** [comment]
**[Charles Karney](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cffk)** commented

[Peter Huffer](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=peter.huffer) Thanks! That worked...

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

[Charles Karney](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cffk) I suppose you don't have defined all the plugin versions you are using..
[Peter Huffer](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=peter.huffer) That would be great so i can create an integration test of it...

**Comment by jira-importer** [comment]
**[Peter Huffer](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=peter.huffer)** commented

[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise) I'm not sure what the best way to upload my example project is. I put up the pom.xml though as well as an empty feature.xml file that should go under a `src/main/feature` directory. If there is a better way to upload the example let me know and I'll do it.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Simplest would be to zip/tar.gz the example project and attach that ;-)...

**Comment by jira-importer** [comment]
**[Peter Huffer](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=peter.huffer)** commented

[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise) ahh duh. Uploaded the zip :)

**Comment by jira-importer** [comment]
**[Guillermo Fernandes](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gsfernandes)** commented

Hi [Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise),

We are having the same issue, seems that 3.0.0-M1 has changed its behaviour when installing a project without main artifact but with attachments.

maven-installer-plugin 2.5.2 was just logging an info message on this scenario:

https://github.com/apache/maven-install-plugin/blob/maven-install-plugin-2.5.2/src/main/java/org/apache/maven/plugin/install/InstallMojo.java#L203

```java
getLog().info( "No primary artifact to install, installing attached artifacts instead." );
```

3.0.0-M1 instead delegates this logic to maven-artifact-transfer which throws an exception on this case:

https://github.com/apache/maven-artifact-transfer/blob/maven-artifact-transfer-0.10.0/src/main/java/org/apache/maven/shared/transfer/project/install/internal/DefaultProjectInstaller.java#L116

I know that would make sense to fail on this scenarios but I'm wondering if this intentional or just you were not aware that this will break builds where plugins declare a custom packaging and are not setting the main artifact to the project.

Regards,

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

So maybe I misunderstand a thing but based on the [docs of karaf-maven-plugin](https://karaf.apache.org/manual/latest/#_maven_plugin)

> The feature packaging verifies a features.xml descriptor using the karaf:verify goal.

This means for me calling `install` on such project is not correct. It does not package/prepare a package like information...

**Comment by jira-importer** [comment]
**[Peter Huffer](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=peter.huffer)** commented

[Guillermo Fernandes](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gsfernandes) I updated the ticket title with that info to be more accurate.

**Comment by jira-importer** [comment]
**[Peter Huffer](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=peter.huffer)** commented

[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise) I think the important part is getting the custom packaging and having attachments without a primary artifact in the project based on [Guillermo Fernandes](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gsfernandes)' comment.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

[Guillermo Fernandes](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gsfernandes) You have given the answer yourself.

> I know that would make sense to fail on this scenarios..

I would say a plugin (own packaging) which does not correctly set the main artifact is doing something wrong...

**Comment by jira-importer** [comment]
**[Guillermo Fernandes](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gsfernandes)** commented

I would say that makes sense to break if the plugin defines a packaging and it not setting the main artifact but... it was working before for a long time... 

So, I would either prefer a warning message instead of this exception and let the whole community know that this will be changed from a warning message to an error in 3.0.0 version. 

The thing here is that suddenly we have seen this issue in customer environments as well in our environments when installing artifacts that use our maven plugin (which defines the custom packaging and is not setting the main artifact) as we were not setting a fixed version of maven-install-plugin as soon as 3.0.0-M1 was deployed they started to use this version and got this issue.

Would it make sense as even this is a milestone version to log a warning and once 3.0.0 is released just throw the error?

I'm wondering how many custom maven plugins may have this issue too.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

[Guillermo Fernandes](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gsfernandes) You are right with ...it was working before for a long time...but that is a major version release which can break and will break things otherwise we would never be able to evolve things...

The point with 3.0.0-M1 is: It is released (now in central) and can't go back or being changed or being removed ....only with 3.0.0-M2 etc. or 3.0.0 I can do changes in it...

What I can do is: Release a version 3.0.0-M2 which will produce a WARNING with a Hint becoming an error with final 3.0.0 release? But usually WARNINGs are ignored...

Furthermore it shows how many people do not correctly pin the version of their plugins...
Apart from that I fear that many custom plugins are doing this wrong...

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Some more hints on the current behaviour: see MINSTALL-118

**Comment by jira-importer** [comment]
**[Guillermo Fernandes](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gsfernandes)** commented

Ok, I understood. Will change the plugin on our side to set the main artifact file. Thank you!

**Comment by jira-importer** [comment]
**[Peter Huffer](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=peter.huffer)** commented

Thanks for the quick responses. I am okay with closing this ticket as won't do and defining the primary artifacts in our plugin's configuration, as well as locking down our plugin version.

**Comment by jira-importer** [comment]
**[zosrothko](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=zosrothko)** commented

Hi

Got the same issue with the izpack-maven-plugin which is running fine with maven-install-plugin:2.5.2 but fails with 3.0.0-M1 as

[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-install-plugin:3.0.0-M1:install (default-install) on project installer-izpack: NoFileAssignedException: The packaging for this project did not assign a file to the build artifact -> [Help 1]

How can this be fixed otherwise to rollback to the maven-install-plugin:2.5.2 ?

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

The problem is located in the izpack-maven-plugin which needs to be fixed...you can configure your build to use an older version...

**Comment by jira-importer** [comment]
**[zosrothko](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=zosrothko)** commented

What should be the fix in the izpaack-maven-plugin?

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

See details in  MINSTALL-118 as mentioned above...

**Comment by jira-importer** [comment]
**[Robert Andersson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=roband)** commented

What about this use-case:

https://stackoverflow.com/questions/352612/how-to-get-maven-to-run-warexploded-but-not-warwar

 

If I understand the discussion above this simple solution will no longer work:

https://stackoverflow.com/a/11134940

 

Instead something more complicated along this line is needed:

https://stackoverflow.com/a/1449740

 

Is the extra complexity for this simple use-case expected and desired?

**Comment by jira-importer** [comment]
**[Leonard Brünings](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=leonard84)** commented

I also ran into this problem. I've created a bom plugin, that has a custom packaging `bom` to create a bill-of-materials pom for the multimodule it is part of.

It works by generating a new pom and setting it via `project.setPomFile` unfortunately I have to set the project packaging to `pom` afterwards, since the `install`/`deploy` plugins have hardcoded logic for `pom` packing that cannot be achieved otherwise.

**Comment by jira-importer** [comment]
**[Andreas Höhmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hoehmi)** commented

```
<profile>
  <id>ca01</id>
  <build>
    <pluginManagement>
      <plugins>
        <!-- Override version for maven-install-plugin because there is a bug in 3.0.0-M1 preventing installing of ca01-war's -->
        <!-- See: https://issues.apache.org/jira/browse/MINSTALL-151 -->
        <plugin>
          <artifactId>maven-install-plugin</artifactId>
          <version>2.5.2</version>
        </plugin>
        <!-- Override version for maven-deploy-plugin because there is a similar bug in 3.0.0-M1 like the install-plugin -->
        <plugin>
          <artifactId>maven-deploy-plugin</artifactId>
          <version>2.8.2</version>
        </plugin>
      </plugins>
    </pluginManagement>
    <plugins>
      <plugin>
        <artifactId>maven-jar-plugin</artifactId>
        <configuration>
          <excludes>
            <exlude>logback.xml</exlude>
            <exlude>**/.gitignore</exlude>
            <exlude>.gitignore</exlude>
          </excludes>
        </configuration>
      </plugin>
      <plugin>
        <artifactId>maven-war-plugin</artifactId>
        <configuration>
          <classifier>ca01</classifier>
          <archiveClasses>true</archiveClasses>
          <attachClasses>true</attachClasses>
          <classesClassifier>ca01-classes</classesClassifier>
              <webResources>
                <resource>
                  <directory>${project.build.directory}</directory>
                  <includes>
                    <include>jetty-env.xml</include>
                  </includes>
                  <targetPath>WEB-INF</targetPath>
                </resource>
          </webResources>
          <packagingExcludes>
                WEB-INF/classes,
                WEB-INF/lib/commons-logging*.jar,
                WEB-INF/lib/log4j*.jar,
                WEB-INF/lib/jcl-over-slf4j*,
                WEB-INF/lib/jul-to-slf4j*.jar,
                WEB-INF/lib/slf4j*.jar,
                WEB-INF/lib/logback*.jar,
                WEB-INF/lib/el-impl*.jar,
                WEB-INF/jetty-web.xml,
                WEB-INF/.faces-config.xml.jsfdia,
                **/.gitignore
          </packagingExcludes>
        </configuration>
      </plugin>
      <plugin>
        <groupId>org.codehaus.mojo</groupId>
        <artifactId>build-helper-maven-plugin</artifactId>
        <executions>
          <execution>
            <id>add-ca01-source</id>
            <phase>generate-sources</phase>
            <goals>
              <goal>add-source</goal>
            </goals>
            <configuration>
              <sources>
                <source>${basedir}/src/main/ca01/java/</source>
              </sources>
            </configuration>
          </execution>
          <execution>
            <id>add-ca01-resources</id>
            <phase>generate-resources</phase>
            <goals>
              <goal>add-resource</goal>
            </goals>
            <configuration>
              <resources>
                <resource>
                  <directory>${basedir}/src/main/ca01/resources/</directory>
                </resource>
              </resources>
            </configuration>
          </execution>
          <execution>
            <id>add-ca01-test-source</id>
            <phase>generate-test-sources</phase>
            <goals>
              <goal>add-test-source</goal>
            </goals>
            <configuration>
              <sources>
                <source>${basedir}/src/test/ca01/java/</source>
              </sources>
            </configuration>
          </execution>
          <execution>
            <id>add-ca01-test-resources</id>
            <phase>generate-test-resources</phase>
            <goals>
              <goal>add-test-resource</goal>
            </goals>
            <configuration>
              <resources>
                <resource>
                  <directory>${basedir}/src/test/ca01/resources/</directory>
                </resource>
              </resources>
            </configuration>
          </execution>
        </executions>
      </plugin>
    </plugins>
  </build>
</profile>
```

With that we are able to build the same web-application in 2 different "flavors"
- online version
- and "ca01 version" -> a kind of offline version

(?) how should I configure install/deploy to archive the same results as with the "old" maven plugins?

Here is what I run ...

```
$ mvn clean install -nsu -DskipTests -Pca01

[INFO]
[INFO] --- maven-jar-plugin:3.1.1:test-jar (default) @ cad-webservice ---
[INFO] Building jar: D:\Dev\git\coc\goos\cad-webservice\target\spicecad-tests.jar
[INFO]
[INFO] --- maven-surefire-plugin:3.0.0-M3:test (default-test) @ cad-webservice ---
[INFO] Tests are skipped.
[INFO]
[INFO] --- git-commit-id-plugin:2.2.6:revision (default) @ cad-webservice ---
[INFO]
[INFO] --- maven-war-plugin:3.2.2:war (default-war) @ cad-webservice ---
[INFO] Packaging webapp
[INFO] Assembling webapp [cad-webservice] in [D:\Dev\git\coc\goos\cad-webservice\target\spicecad]
[INFO] Processing war project
[INFO] Copying webapp webResources [D:\Dev\git\coc\goos\cad-webservice\target] to [D:\Dev\git\coc\goos\cad-webservice\target\spicecad]
[INFO] Copying webapp resources [D:\Dev\git\coc\goos\cad-webservice\src\main\webapp]
[INFO] Building jar: D:\Dev\git\coc\goos\cad-webservice\target\spicecad\WEB-INF\lib\cad-webservice-2.2019.02.0-SNAPSHOT.jar
[INFO] Webapp assembled in [2106 msecs]
[INFO] Building war: D:\Dev\git\coc\goos\cad-webservice\target\spicecad-ca01.war
[INFO]
[INFO] --- maven-source-plugin:3.0.1:jar-no-fork (attach-sources) @ cad-webservice ---
[INFO] Building jar: D:\Dev\git\coc\goos\cad-webservice\target\spicecad-sources.jar
[INFO]
[INFO] --- maven-source-plugin:3.0.1:test-jar-no-fork (attach-sources) @ cad-webservice ---
[INFO] Building jar: D:\Dev\git\coc\goos\cad-webservice\target\spicecad-test-sources.jar
[INFO]
[INFO] --- maven-javadoc-plugin:3.0.1:jar (attach-javadocs) @ cad-webservice ---
[INFO] Skipping javadoc generation
[INFO]
[INFO] --- maven-javadoc-plugin:3.0.1:test-jar (attach-javadocs) @ cad-webservice ---
[INFO] Skipping javadoc generation
[INFO]
[INFO] --- maven-install-plugin:3.0.0-M1:install (default-install) @ cad-webservice ---
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 43.496 s
[INFO] Finished at: 2019-01-11T09:45:25+01:00
[INFO] Final Memory: 93M/1238M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-install-plugin:3.0.0-M1:install (default-install) on project cad-webservice: NoFileAssignedException: The packaging plugin for this project did not assign a main file to the project but it has attachments. Change packaging to 'pom'. -> [Help 1]


$ ls -la target/
total 93960
drwxr-xr-x 1 hoehmann 1049089        0 Jan 11 09:45 classes/
drwxr-xr-x 1 hoehmann 1049089        0 Jan 11 09:44 generated-sources/
drwxr-xr-x 1 hoehmann 1049089        0 Jan 11 09:45 generated-test-sources/
-rw-r--r-- 1 hoehmann 1049089     7927 Jan 11 09:44 jetty-env.xml
drwxr-xr-x 1 hoehmann 1049089        0 Jan 11 09:45 maven-archiver/
drwxr-xr-x 1 hoehmann 1049089        0 Jan 11 09:45 spicecad/
-rw-r--r-- 1 hoehmann 1049089 94863803 Jan 11 09:45 spicecad-ca01.war
-rw-r--r-- 1 hoehmann 1049089   251161 Jan 11 09:45 spicecad-classes.jar
-rw-r--r-- 1 hoehmann 1049089   161729 Jan 11 09:45 spicecad-sources.jar
-rw-r--r-- 1 hoehmann 1049089   473024 Jan 11 09:45 spicecad-tests.jar
-rw-r--r-- 1 hoehmann 1049089   431800 Jan 11 09:45 spicecad-test-sources.jar
drwxr-xr-x 1 hoehmann 1049089        0 Jan 11 09:45 test-classes/
-rw-r--r-- 1 hoehmann 1049089     5057 Jan 11 09:44 tomcat-context.xml
```

And here the "normal" build which works:

```
$mvn clean install -nsu -DskipTests

[INFO]
[INFO] --- maven-surefire-plugin:3.0.0-M3:test (default-test) @ cad-webservice ---
[INFO] Tests are skipped.
[INFO]
[INFO] --- git-commit-id-plugin:2.2.6:revision (default) @ cad-webservice ---
[INFO]
[INFO] --- maven-war-plugin:3.2.2:war (default-war) @ cad-webservice ---
[INFO] Packaging webapp
[INFO] Assembling webapp [cad-webservice] in [D:\Dev\git\coc\goos\cad-webservice\target\spicecad]
[INFO] Processing war project
[INFO] Copying webapp resources [D:\Dev\git\coc\goos\cad-webservice\src\main\webapp]
[INFO] Webapp assembled in [1536 msecs]
[INFO] Building war: D:\Dev\git\coc\goos\cad-webservice\target\spicecad.war
[INFO]
[INFO] --- maven-source-plugin:3.0.1:jar-no-fork (attach-sources) @ cad-webservice ---
[INFO] Building jar: D:\Dev\git\coc\goos\cad-webservice\target\spicecad-sources.jar
[INFO]
[INFO] --- maven-source-plugin:3.0.1:test-jar-no-fork (attach-sources) @ cad-webservice ---
[INFO] Building jar: D:\Dev\git\coc\goos\cad-webservice\target\spicecad-test-sources.jar
[INFO]
[INFO] --- maven-javadoc-plugin:3.0.1:jar (attach-javadocs) @ cad-webservice ---
[INFO] Skipping javadoc generation
[INFO]
[INFO] --- maven-javadoc-plugin:3.0.1:test-jar (attach-javadocs) @ cad-webservice ---
[INFO] Skipping javadoc generation
[INFO]
[INFO] --- maven-install-plugin:3.0.0-M1:install (default-install) @ cad-webservice ---
[INFO] Installing D:\Dev\git\coc\goos\cad-webservice\target\spicecad.war to D:\Dev\ApacheMaven\repository\com\siemens\spice\cad-webservice\2.2019.02.0-SNAPSHOT\cad-webservice-2.2019.02.0-SNAPSHOT.war
[INFO] Installing D:\Dev\git\coc\goos\cad-webservice\pom.xml to D:\Dev\ApacheMaven\repository\com\siemens\spice\cad-webservice\2.2019.02.0-SNAPSHOT\cad-webservice-2.2019.02.0-SNAPSHOT.pom
[INFO] Installing D:\Dev\git\coc\goos\cad-webservice\target\spicecad-sources.jar to D:\Dev\ApacheMaven\repository\com\siemens\spice\cad-webservice\2.2019.02.0-SNAPSHOT\cad-webservice-2.2019.02.0-SNAPSHOT-sources.jar
[INFO] Installing D:\Dev\git\coc\goos\cad-webservice\target\spicecad-test-sources.jar to D:\Dev\ApacheMaven\repository\com\siemens\spice\cad-webservice\2.2019.02.0-SNAPSHOT\cad-webservice-2.2019.02.0-SNAPSHOT-test-sources.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 22.407 s
[INFO] Finished at: 2019-01-11T09:48:40+01:00
[INFO] Final Memory: 94M/1394M
[INFO] ------------------------------------------------------------------------

$ ls -la target/
total 93996
drwxr-xr-x 1 hoehmann 1049089        0 Jan 11 09:48 classes/
drwxr-xr-x 1 hoehmann 1049089        0 Jan 11 09:48 generated-sources/
drwxr-xr-x 1 hoehmann 1049089        0 Jan 11 09:48 generated-test-sources/
-rw-r--r-- 1 hoehmann 1049089     8100 Jan 11 09:48 jetty-env.xml
drwxr-xr-x 1 hoehmann 1049089        0 Jan 11 09:48 maven-archiver/
drwxr-xr-x 1 hoehmann 1049089        0 Jan 11 09:48 spicecad/
-rw-r--r-- 1 hoehmann 1049089 95631598 Jan 11 09:48 spicecad.war
-rw-r--r-- 1 hoehmann 1049089   161729 Jan 11 09:48 spicecad-sources.jar
-rw-r--r-- 1 hoehmann 1049089   431800 Jan 11 09:48 spicecad-test-sources.jar
drwxr-xr-x 1 hoehmann 1049089        0 Jan 11 09:48 test-classes/
-rw-r--r-- 1 hoehmann 1049089     5225 Jan 11 09:48 tomcat-context.xml
```

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

[Leonard Brünings](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=leonard84) Can you please give a link to your pom plugin ?
[Robert Andersson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=roband) have you tried the solutions if they work or not?
[Andreas Höhmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hoehmi) Based on your configuration the `normal` build runs without any issue cause the maven-war-plugin sets correctly a main artifact. Based on your profile which configures maven-war-plugin the way to produce a file which has a classifier but without having a main artifact. An artifact with a classifier can only exist if a main artifact exists..which is not the case based on the given configuration. So the solution is to correctly run without a profile to produce always correctly the main artifact (via usual maven-war-plugin configuration) and a supplemental artifact via classifier or even better create a separate module which contains the supplemental/changed code and different dependencies which would result in removing the profile, the supplemental configuration of maven-war-plugin, packageExcludes and removing the need for build-helper-maven-plugin as well.

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

[Domagoj Cosic](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=domagoj) can you provide a reproducer? Also, what is your definition of "main artifact", as it is per defitionem "artifact without classifier"

**Comment by jira-importer** [comment]
**[Peter De Maeyer](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=peterdm)** commented

My use case: some of my submodules in my multi-module project only produce artifacts of type "test-jar". In other words, they don't have produce a main `my-artifact-1.0.0.jar`, but they _do_ produce a `my-artifact-1.0.0-tests.jar`. In my parent POM I have a section which fits for any module that produces either or both a main or test artifact, as supported by the default `maven-jar-plugin`:

```xml
<plugin>
    <artifactId>maven-jar-plugin</artifactId>
    <executions>
        <execution>
            <!-- The ID must match the ID of the default goal, because we want to reconfigure it. -->
            <id>default-jar</id>
            <goals>
                <goal>jar</goal>
            </goals>
            <!-- Don't produce an artifact of type "jar" if main/src is empty. -->
            <configuration>
                <skipIfEmpty>true</skipIfEmpty>
            </configuration>
        </execution>
        <execution>
            <goals>
                <goal>test-jar</goal>
            </goals>
            <!-- Don't produce an artifact of type "test-jar" if test/src is empty. -->
            <configuration>
                <skipIfEmpty>true</skipIfEmpty>
            </configuration>
        </execution>
    </executions>
</plugin>
```

`maven-install-plugin` fails with the above "The packaging plugin for this project did not assign a main file to the project but it has attachments. Change packaging to 'pom'."

The advice is wrong, I can't change packaging to 'pom' because then it won't compile nor run my tests.

**Comment by jira-importer** [comment]
**[Claudio Clemens](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER299313)** commented

Hi, I have also this problem when I create artifacts with classifiers. I don't know if the problem is in the maven-install-plugin or in the maven-jar-plugin.

I could work around the problem, if disable the default "install" goal, and instead call a "install-file" goal. Take a look at mvn-install-fail.zip where the problem is demonstrated in "failing", and the "fix" in "workaround".

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Version 3.1.1 (just released) offers "escape hatch" in form of https://maven.apache.org/plugins/maven-install-plugin/install-mojo.html#allowIncompleteProjects
